### PR TITLE
Improve #head builds

### PR DESCRIPTION
**Pull request by genotrance**
- Use `git ls-remote` to get latest commit hash for csources and #head - helps get unique file names
- Revert to current behavior if git isn't available
- Delete `csources` directory on successful build, similar to `c_code`

### CODE REVIEW

File: src/choosenim/download.nim [hunk 1]
```diff
@@ -205,15 +206,18 @@ proc needsDownload(params: CliParams, downloadUrl: string,
...
+      reference =
+        case normalize($version)
+        of "#head":
+          archive
+        else:
+          ($version)[1 .. ^1]
     display("Downloading", "Nim $1 from $2" % [reference, "GitHub"],
             priority = HighPriority)
-    let url = githubUrl % reference
+    let url = githubUrl & (dlArchive % reference)
```
Comment (dom96) [inline src/choosenim/download.nim]: consider using ``parseUri(...) / (dlArchive % reference)`` for less error prone concat of urls. You might even be able to put the `parseUri` into the `const`.

File: src/choosenim/download.nim [hunk 2]
```diff
@@ -252,12 +256,17 @@ proc download*(version: Version, params: CliParams): string =
                        $version)
 
 proc downloadCSources*(params: CliParams): string =
+  let
+    commit = getLatestCommit(csourcesUrl, "master")
+    archive = if commit.len != 0: commit else: "master"
+    csourcesDLUrl = csourcesUrl & (dlArchive % archive)
```
Comment (dom96) [inline src/choosenim/download.nim]: Nit:

```suggestion
    csourcesArchiveUrl = csourcesUrl & (dlArchive % archive)
```

File: src/choosenim/builder.nim [hunk 3]
```diff
@@ -47,6 +47,9 @@ proc buildCompiler(params: CliParams) =
       display("Building", "Nim", priority = HighPriority)
       doCmdRaw("./koch boot -d:release")
 
+    # Not removing csources archive since repo is now locked
+    # discard tryRemoveFile(path)
```
Comment (dom96) [inline src/choosenim/builder.nim]: Not entirely sure what this comment means, but if this is no longer necessary maybe just remove it?

File: src/choosenim/download.nim (hunk 1 above)
Comment (genotrance) [reply src/choosenim/download.nim]: I had tried `/` but it changed `https://xxx` into `https:/xxx`. Why do we need to parse the URL here?

File: src/choosenim/builder.nim (hunk 3 above)
Comment (genotrance) [reply src/choosenim/builder.nim]: I can remove it, was just a reminder for the future.

File: src/choosenim/download.nim (hunk 1 above)
Comment (dom96) [reply src/choosenim/download.nim]: because then you'll use the Uri overload for `/` which will work correctly.

**Comment by dom96** [comment]
> The build still goes into ~/.choosenim/toolchains/nim-#head. If you run choosenim #head again, it will just switch to it. Whereas if you run choosenim update #head, it will redownload and rebuild.

That doesn't seem to follow what the code shows. My understanding is that builds will go into `~/.choosenim/toolchains/nim-<commithash>` where `<commithash>` is the latest commit hash.

Can you explain precisely where everything ends up with each command that is run? ``choosenim #head`` then ``choosenim update #head``.

File: src/choosenim/download.nim (hunk 1 above)
Comment (genotrance) [reply src/choosenim/download.nim]: Done

File: src/choosenim/download.nim (hunk 2 above)
Comment (genotrance) [reply src/choosenim/download.nim]: Done

File: src/choosenim/builder.nim (hunk 3 above)
Comment (genotrance) [reply src/choosenim/builder.nim]: Done

**Comment by timotheecour** [comment]
> Delete csources directory on successful build, similar to c_code

won't that cause more network traffic (to re-download whenever rebuild needed) ? network traffic matters most than disk space; all we need would be to cache each unique csources release (at most the number of git tags = 3 instances total)

**Comment by genotrance** [comment]
The c_code / csources directory being deleted is within `~/.choosenim/toolchains/nim-xxx`. A nim source archive is only 12MB in size (xz) but extracts to 700+MB, 650 of which is this c_code directory. If you install 4-5 Nim versions with choosenim, you end up with several gigs of data. Further, all recent Nim versions are binary installs so the csources are not even needed.

As for csources that are downloaded as part of #head or commit builds, that is cached and stored in downloads and reused since csources is now locked.

**Comment by timotheecour** [comment]
ok convincing arguments :)

**Pull request merged**